### PR TITLE
Modernize and use `.fork()` and `.greenlet()`

### PR DESCRIPTION
**Pull request by dlecocq**
Previously, we used to declare whether we were going to use `fork` or `greenlet` when creating the server. It's kind of annoying, it's inflexible, unnecessary, and was confusing.

Instead... introduce two different context managers: `fork` and `greenlet`

@b4hand

### CODE REVIEW

File: test.py [hunk 1]
```diff
@@ -114,25 +119,71 @@ def test_encoding(self):
...
+            self.assertEqual(req.status_code, 200)
+
+    def test_fork_early_exit(self):
+        '''Raises an exception if the child process exits early.'''
+        server = asis.Server('test', port=8081, server='gevent')
+        with mock.patch.object(server, 'run', side_effect=ValueError('kaboom')):
+            with self.assertRaises(RuntimeError):
+                with server.fork():
+                    pass
+
```
Comment (b4hand) [inline test.py]: Lots of trailing blank lines?

File: asis.py [hunk 2]
```diff
@@ -39,8 +42,7 @@
 logger.setLevel(logging.WARNING)
 
 
-class Server(object):
-    '''Regular expression for matching found headers'''
```
Comment (b4hand) [inline asis.py]: Did you mean to delete this docstring? I think it was intended to apply to `headerMatch` rather than the class.

File: asis.py [hunk 3]
```diff
@@ -146,7 +132,7 @@ def read(self, path):
                 logger.debug('Encoding to character set: %s' % charset)
                 content = content.decode('utf-8').encode(charset)
                 # Update the Content-Length if one was included
-                if response.headers.get('Content-Length'):
+                if response.headers.get('Content-Length') is not None:
```
Comment (b4hand) [inline asis.py]: This change seems unrelated to this PR.

File: asis.py [hunk 4]
```diff
@@ -178,55 +164,86 @@ def handle(self, path):
...
-        '''Actually run the server, whether it's in a separate process, or
-        blocking or in a green thread'''
-        try:
-            run(self.app, host='', port=self.port, server=self.server)
-        except KeyboardInterrupt:
-            # Finished
-            pass
+
+class Server(object):
+    '''Regular expression for matching found headers'''
```
Comment (b4hand) [inline asis.py]: Looks like this docstring got incorrectly moved.

File: asis.py (hunk 2 above)
Comment (dlecocq) [inline asis.py]: D'oh. Fixed.

File: asis.py (hunk 3 above)
Comment (dlecocq) [inline asis.py]: Yep

File: asis.py (hunk 4 above)
Comment (dlecocq) [inline asis.py]: Updated

**Comment by dlecocq** [comment]
Reupped

File: test.py [hunk 5]
```diff
@@ -114,25 +119,55 @@ def test_encoding(self):
...
-        url = 'http://localhost:8080/basic/basic.asis'
-        # Shouldn't be able to fetch
-        self.assertRaises(requests.exceptions.ConnectionError,
-            requests.get, url)
-        with asis.Server('test', port=8080, server='gevent'):
-            # Make sure we can get requests now...
-            req = requests.get(url)
+    def test_charset_no_length(self):
+        '''When a charset is used, but a content-length is absent, none is provided.'''
+        # Just exercise this branch
```
Comment (b4hand) [inline test.py]: Is it that complicated to add an assertion?

File: test.py [hunk 6]
```diff
@@ -114,25 +119,55 @@ def test_encoding(self):
...
+        requests.get(self.base + 'basic/charset-no-length.asis')
+
+    def test_encoding_no_length(self):
+        '''When an encoding is used, but a content-length is absent, none is provided.'''
+        # Just exercise this branch
```
Comment (b4hand) [inline test.py]: Same here.

**Comment by b4hand** [comment]
Except for the last nits about assertion-less tests, LGTM.

File: test.py (hunk 5 above)
Comment (dlecocq) [inline test.py]: So, if no `Content-Length` header is provided in this case we don't want to create one. But normally if one is provided, then we adjust it according to its new compressed / encoded length.

However, it seems `bottle` is adding `Content-Length` anyways, so I don't know that there's really a meaningful assertion to be done here.

Comment (dlecocq) [inline test.py]: Turns out that `bottle` _really_ doesn't want you to return no `Content-Length`.
